Readme.md
# ChatGPT API 

This is a simple Python client/server using GraphQL to interact with the OpenAI ChatGPT API. It demonstrates how to send messages to the API and receive responses in a conversation-like manner. The server also includes functionality to mask sensitive information and process Japanese text.

## Requirements

- Python 3.6 or higher
- Install the required packages by running: `pip install -r requirements.txt`

## Installation

Follow these steps to set up the ChatGPT API server:

1. Ensure you have Python 3.6 or higher installed on your system.

2. Clone the repository or download the source files to your local machine.

3. Navigate to the project directory in your terminal.

4. Install the required packages by running:

```bash
pip install -r requirements.txt
```

5. Download the necessary spaCy language model


That's it. Now you can follow the usage instructions in the README to start the server and interact with the ChatGPT API.



## Usage

1. Set your OpenAI API key as an environment variable:

```bash
export OPENAI_API_KEY="your-api-key-here"
```

2. Start the GraphQL server by running,
```bash
python graphQL_server.py
```
3. Run the graphQL_client.py script with the following arguments:

* --system: The system message to start the conversation (optional).
* --assistant: The assistant message (optional).
* --user: The user message (required).

```bash
python graphQL_client.py --system "be my English teacher" --assistant "I will correct your English" --user "Your text message here"
```

## Features
### Masking Sensitive Information
The client can mask sensitive information like names, phone numbers, credit card numbers, and expiration dates in the user's messages. To enable this feature, apply the mask_sensitive_data function to the user message.



## Supported ChatGPT Models
As of now, the following ChatGPT models can be used with this server:

* gpt-4
* gpt-4-0314
* gpt-4-32k
* gpt-4-32k-0314
* gpt-3.5-turbo
* gpt-3.5-turbo-0301

You can modify the model_name variable in the graphQL_server.py script to use a different model.


### Handling Japanese Text
The client can handle Japanese text by converting it to Unicode escape characters before sending it to the API. To use this feature, apply the unicode_escape function to the user message, and the unicode_unescape function to the API response.

## Customization

Feel free to modify the graphQL_client.py script to add your own filters or implement additional functionality as needed.

## Known Issues
Some users have reported issues with sending Japanese text to the API. Please ensure that the OpenAI API supports Unicode escape characters before using the Japanese text handling feature.

## Lincense
MIT


history.txt
user: Hello
assistant: Hi there!
user: How are you?
assistant: I'm doing well, thanks for asking. How about you?
user: I'm good too.
assistant: That's great to hear.




test_app.py
import pytest
from graphene.test import Client
from graphQL_server import app, schema, anonymize_text, contains_confidential_information, read_keywords_from_file, load_spacy_model


@pytest.fixture
def client():
    app.config['TESTING'] = True
    with app.test_client() as client:
        yield client

def test_authors(client):
    query = """
    query {
        authors {
            id
            name
            country
        }
    }
    """
    client = Client(schema)
    response = client.execute(query)
    assert response['data']['authors'] is not None
    assert len(response['data']['authors']) == 2

def test_books(client):
    query = """
    query {
        books {
            id
            title
            author {
                id
                name
                country
            }
        }
    }
    """
    client = Client(schema)
    response = client.execute(query)
    assert response['data']['books'] is not None
    assert len(response['data']['books']) == 3

def test_system_messages(client):
    query = """
    query {
        systemMessages {
            content
        }
    }
    """
    client = Client(schema)
    response = client.execute(query)
    assert response['data']['systemMessages'] is not None
    assert len(response['data']['systemMessages']) == 1


# def test_assistant_messages(client):
#     query = """
#     query {
#         assistantMessages {
#             content
#         }
#     }
#     """
#     client = Client(schema)
#     response = client.execute(query)
#     filtered_messages = response['data']['assistantMessages']
#     print(f"Filtered messages: {filtered_messages}")
#     assert filtered_messages is not None
#     assert len(filtered_messages) == 1


def test_read_keywords_from_file():
    keywords = read_keywords_from_file("keywords.txt")
    assert len(keywords) > 0
    assert "secret" in keywords


def test_contains_confidential_information():
    confidential_message = "This message contains a secret keyword."
    non_confidential_message = "This message does not have any sensitive data."
    contains_confidential, keyword = contains_confidential_information(confidential_message)
    assert contains_confidential == True
    contains_confidential, keyword = contains_confidential_information(non_confidential_message)
    assert contains_confidential == False, f"False positive caused by keyword: {keyword}"



def test_anonymize_text():
    load_spacy_model()
    original_text = "田中太郎と鈴木花子は東京から大阪に引っ越しました。"
    expected_anonymized_text = "{name}と{name}は{place}から{place}に引っ越しました。"
    anonymized_text = anonymize_text(original_text)
    assert anonymized_text == expected_anonymized_text


if __name__ == "__main__":
    pytest.main()


graphQL_client.py
import os
import openai
import argparse
import requests
import json

openai.api_key = os.getenv("OPENAI_API_KEY")

def send_graphql_request(query):
    url = "http://localhost:5000/graphql"
    headers = {
        "Content-Type": "application/json",
    }
    data = json.dumps({"query": query})
    response = requests.post(url, headers=headers, data=data)

    if response.status_code == 200:
        return json.loads(response.text)
    else:
        raise Exception(f"GraphQL request failed with status code {response.status_code}")


def parse_arguments():
    parser = argparse.ArgumentParser(description="Chat with OpenAI models using GraphQL server")
    parser.add_argument("--model", type=str, default="gpt-4", help="Model name")
    return parser.parse_args()




# def main():
#     args = parse_arguments()
#     model_name = args.model
#     if model_name is None:
#         print("Please specify a model name")
#         return

#     # GraphQL query to fetch system and assistant messages
#     system_and_assistant_query = """
#     query {
#         system_messages {
#             content
#         }
#         assistant_messages {
#             content
#         }
#     }
#     """

#     # GraphQL query to fetch book data
#     book_data_query = """
#     query {
#         books {
#             title
#             author {
#                 name
#                 country
#             }
#         }
#     }
#     """

#     system_and_assistant_response = send_graphql_request(system_and_assistant_query)
#     book_data_response = send_graphql_request(book_data_query)

#     if 'errors' in system_and_assistant_response:
#         print("Error fetching data from GraphQL API:")
#         print(system_and_assistant_response['errors'])
#         return

#     if 'errors' in book_data_response:
#         print("Error fetching data from GraphQL API:")
#         print(book_data_response['errors'])
#         return

#     system_messages = system_and_assistant_response['data']['system_messages']
#     assistant_messages = system_and_assistant_response['data']['assistant_messages']
#     book_data = book_data_response['data']['books'][0]
#     book_title = book_data['title']
#     author_name = book_data['author']['name']
#     author_country = book_data['author']['country']
    
#     user_message = f"Tell me more about the book '{book_title}' written by {author_name} from {author_country}."

#     messages = []
#     for msg in system_messages:
#         messages.append({"role": "system", "content": msg['content']})
#     for msg in assistant_messages:
#         messages.append({"role": "assistant", "content": msg['content']})
#     messages.append({"role": "user", "content": user_message})

#     response = chat_with_models(model_name, messages)
#     if response is not None:
#         print(f"{model_name}'s response: {response}")


def main():
    # GraphQL query to fetch system and assistant messages
    system_and_assistant_query = """
    query {
        systemMessages {
            content
        }
        assistantMessages {
            content
        }
    }
    """

    system_and_assistant_response = send_graphql_request(system_and_assistant_query)

    if 'errors' in system_and_assistant_response:
        print("Error fetching data from GraphQL API:")
        print(system_and_assistant_response['errors'])
    else:
        print("System and Assistant Messages:")
        print(system_and_assistant_response)


if __name__ == "__main__":
    main()


graphQL_server.py
import os
import random
from flask import Flask
from flask_graphql import GraphQLView
import graphene
import sqlite3
import re
import spacy

# graphQL_server.py
def load_spacy_model():
    global nlp
    nlp = spacy.load("ja_core_news_sm")

class Author(graphene.ObjectType):
    id = graphene.Int()
    name = graphene.String()
    country = graphene.String()

class Book(graphene.ObjectType):
    id = graphene.Int()
    title = graphene.String()
    author = graphene.Field(Author)

# Sample data
authors = [
    Author(id=1, name='George Orwell', country='United Kingdom'),
    Author(id=2, name='Aldous Huxley', country='United Kingdom'),
]

books = [
    Book(id=1, title='1984', author=authors[0]),
    Book(id=2, title='Animal Farm', author=authors[0]),
    Book(id=3, title='Brave New World', author=authors[1]),
]

class SystemMessage(graphene.ObjectType):
    content = graphene.String()

class AssistantMessage(graphene.ObjectType):
    content = graphene.String()


system_messages = [
    SystemMessage(content="You are now chatting with GPT-4."),
]

assistant_messages = [
    AssistantMessage(content="Hello! I'm GPT-4, and I'm here to help you."),
]



def read_keywords_from_file(file_path):
    with open(file_path, 'r', encoding='utf-8') as file:
        keywords = [line.strip() for line in file.readlines()]
    return keywords

confidential_keywords = read_keywords_from_file(os.path.join(os.path.dirname(__file__), 'keywords.txt'))

def find_names_and_places(text):
    doc = nlp(text)
    names = [ent.text for ent in doc.ents if ent.label_ == "PERSON"]
    places = [ent.text for ent in doc.ents if ent.label_ == "GPE"]

    return names, places

def anonymize_text(text):
    names, places = find_names_and_places(text)
    name_placeholder = "{name}"
    place_placeholder = "{place}"
    
    for name in names:
        text = re.sub(re.escape(name), name_placeholder, text, flags=re.IGNORECASE | re.UNICODE)
    
    for place in places:
        text = re.sub(re.escape(place), place_placeholder, text, flags=re.IGNORECASE | re.UNICODE)

    return text

def contains_confidential_information(message):
    for keyword in confidential_keywords:
        if keyword.lower() in message.lower():
            print(f"Keyword '{keyword}' found in message: '{message}'")
            return True, keyword
    return False, None

# def contains_confidential_information(message):
#     for keyword in confidential_keywords:
#         if keyword.lower() in message.lower():
#             return True
#     return False


class Query(graphene.ObjectType):
    authors = graphene.List(Author)
    books = graphene.List(Book)
    system_messages = graphene.List(SystemMessage)
    print(AssistantMessage)
    assistant_messages = graphene.List(AssistantMessage)

    def resolve_authors(self, info):
        return authors

    def resolve_books(self, info):
        return books

    
    def resolve_system_messages(self, info):
        # Filter out messages containing confidential information
        filtered_messages = []
        for msg in system_messages:
            contains_confidential, keyword = contains_confidential_information(msg.content)
            if not contains_confidential:
                filtered_messages.append(msg)
            else:
                print(f"Filtered out message: {msg.content} | Detected keyword: {keyword}")
        return filtered_messages


    def resolve_assistant_messages(self, info):
        # Filter out messages containing confidential information
        
        filtered_messages = [msg for msg in assistant_messages if not contains_confidential_information(msg.content)]
        return filtered_messages

schema = graphene.Schema(query=Query)

app = Flask(__name__)
app.add_url_rule('/graphql', view_func=GraphQLView.as_view('graphql', schema=schema, graphiql=True))


def read_keywords_from_file(file_path):
    with open(file_path, 'r', encoding='utf-8') as file:
        keywords = [line.strip() for line in file.readlines()]
    return keywords

# Read keywords from the file
confidential_keywords = read_keywords_from_file('keywords.txt')


if __name__ == '__main__':
    app.run()


chatgpt.py
import os
import openai
import argparse

openai.api_key = os.getenv("OPENAI_API_KEY")

# add coment

def chat_with_models(model, messages):
    '''
    Chat with OpenAI models
    Args:
        model (str): Model name
        messages (list): List of messages
    Returns:
        str: Response from the model
    '''
    try:
        chat_models = [
            "gpt-4", "gpt-4-0314", "gpt-4-32k", "gpt-4-32k-0314",
            "gpt-3.5-turbo", "gpt-3.5-turbo-0301"
        ]

        if model in chat_models:
            response = openai.ChatCompletion.create(
                model=model,
                messages=messages
            )
            return response.choices[0].message.content
        else:
            prompt = " ".join([msg["content"] for msg in messages if msg["role"] == "user"])
            response = openai.Completion.create(
                engine=model,
                prompt=prompt,
                max_tokens=150,
                n=1,
                stop=None,
                temperature=0.5
            )
            return response.choices[0].text.strip()

    except openai.OpenAIError as e:
        print(f"Error occurred while connecting to OpenAI API: {e}")
        print("Possible causes:")
        print("1. Invalid API key")
        print("2. Network issues")
        print("3. Invalid model or unsupported model for chat")
        print("4. Issues with the OpenAI API service")
        return None


def parse_arguments():
    parser = argparse.ArgumentParser(description="Chat with OpenAI models")
    parser.add_argument("--model", type=str,default="gpt-3.5-turbo-0301", help="Model name")
    parser.add_argument("--system", nargs="*", default=[], help="System message(s)")
    parser.add_argument("--assistant", nargs="*", default=[], help="Assistant message(s)")
    parser.add_argument("--user", nargs="*", default=[], help="User message(s)")
    return parser.parse_args()


def main():
    args = parse_arguments()
    model_name = args.model

    messages = []
    for msg in args.system:
        messages.append({"role": "system", "content": msg})
    for msg in args.assistant:
        messages.append({"role": "assistant", "content": msg})
    for msg in args.user:
        messages.append({"role": "user", "content": msg})

    response = chat_with_models(model_name, messages)
    if response is not None:
        print(f"{model_name}'s response: {response}")

if __name__ == "__main__":
    main()